core/src/main/java/org/apache/iceberg/DefaultValueParser.java

```diff
@@ -0,0 +1,348 @@
+        generator.writeBoolean((Boolean) defaultValue);
+        break;
+      case INTEGER:
+        Preconditions.checkArgument(
+            defaultValue instanceof Integer, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Integer) defaultValue);
+        break;
+      case LONG:
+        Preconditions.checkArgument(defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Long) defaultValue);
+        break;
+      case FLOAT:
+        Preconditions.checkArgument(defaultValue instanceof Float, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Float) defaultValue);
+        break;
+      case DOUBLE:
+        Preconditions.checkArgument(defaultValue instanceof Double, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeNumber((Double) defaultValue);
+        break;
+      case DATE:
+        Preconditions.checkArgument(
+            defaultValue instanceof Integer, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(DateTimeUtil.daysToIsoDate((Integer) defaultValue));
+        break;
+      case TIME:
+        Preconditions.checkArgument(
+            defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(DateTimeUtil.microsToIsoTime((Long) defaultValue));
+        break;
+      case TIMESTAMP:
+        Preconditions.checkArgument(
+            defaultValue instanceof Long, "Invalid default %s value: %s", type, defaultValue);
+        if (((Types.TimestampType) type).shouldAdjustToUTC()) {
+          generator.writeString(DateTimeUtil.microsToIsoDateTimeTz((Long) defaultValue));
+        } else {
+          generator.writeString(DateTimeUtil.microsToIsoDateTime((Long) defaultValue));
+        }
+        break;
+      case STRING:
+        Preconditions.checkArgument(
+            defaultValue instanceof CharSequence, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(((CharSequence) defaultValue).toString());
+        break;
+      case UUID:
+        Preconditions.checkArgument(defaultValue instanceof UUID, "Invalid default %s value: %s", type, defaultValue);
+        generator.writeString(defaultValue.toString());
+        break;
+      case FIXED:
+        Preconditions.checkArgument(
+            defaultValue instanceof ByteBuffer, "Invalid default %s value: %s", type, defaultValue);
+        String fixedString = BaseEncoding.base16().encode(ByteBuffers.toByteArray(((ByteBuffer) defaultValue)));
+        int valueLength = fixedString.length();
+        int expectedLength = ((Types.FixedType) type).length();
+        Preconditions.checkArgument(valueLength == expectedLength * 2,
+            "Invalid default %s value: %s, incorrect length: %s",
+            type, defaultValue, valueLength);
```

Comment:
Minor: this should check the `remaining` bytes in the incoming buffer rather than checking the final string value. That would avoid a conversion and get the check done as soon as possible.

Reply:
Done